pkg/library/libraryinputresources/resource.go

```diff
@@ -63,7 +65,57 @@ func LenientResourcesFromDirRecursive(location string) ([]*Resource, error) {
 	return currResourceList, errors.Join(errs...)
 }
 
-func ResourceFromFile(location, fileTrimPrefix string) (*Resource, error) {
+func EnsureResourceType(discoveryClient discovery.DiscoveryInterface, resources []*Resource) error {
+	var errs []error
+	gvkToAPIResourceList := map[schema.GroupVersionKind]*v1.APIResourceList{}
+	// Build a cache to avoid repetitive discovery calls
+	for _, resource := range resources {
+		gvk := resource.Content.GetObjectKind().GroupVersionKind()
+		if _, ok := gvkToAPIResourceList[gvk]; !ok {
+			list, err := resourceListForGVK(discoveryClient, gvk)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			gvkToAPIResourceList[gvk] = list
+		}
+		// Find the GVR for the current GVK
+		gvr, err := findGVR(gvkToAPIResourceList[gvk], gvk)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		resource.ResourceType = *gvr
+	}
+	return errors.Join(errs...)
+}
```

Comment:
prefer non-mutating functions.  Let's build this as we go.

Reply:
done